pdg/src/info.rs

```diff
@@ -0,0 +1,515 @@
+/// and neg), as well as whether they are unique.
+/// Uniqueness of node X is determined based on whether there is a node Y which is X's ancestor
+/// through copies, fields, and offsets, also is a node Z's ancestor through copies, fields, offsets,
+/// where the fields are the same between X and Z, and where Z is chronologically between X and X's last descendent.
+/// If such a node exists, X is not unique.
+#[derive(Debug, Eq, PartialEq, Hash, Clone)]
+pub struct NodeInfo {
+    flows_to_mutation: Option<NodeId>,
+    flows_to_load: Option<NodeId>,
+    flows_to_pos_offset: Option<NodeId>,
+    flows_to_neg_offset: Option<NodeId>,
+    non_unique: Option<NodeId>,
+}
+
+//Algorithm for determining non-conflicting:
+//First calculate all of the below from the bottom-up.
+//Then create a tree where children are based on fields only,
+//each node having a list of NodeIds associated with it.
+//Finally, check that each node's starts and ends don't overlap with each other.
+#[derive(Debug, Hash, Clone, Copy)]
+struct GraphTraverseInfo {
+    last_descendent: NodeId,
+    flows_to_load: Option<NodeId>,
+    flows_to_store: Option<NodeId>,
+    flows_to_pos_offset: Option<NodeId>,
+    flows_to_neg_offset: Option<NodeId>,
+}
+
+fn init_traverse_info (n_id: NodeId, n: &Node) -> GraphTraverseInfo {
+    GraphTraverseInfo {
+        last_descendent: n_id,
+        flows_to_store: node_does_mutation(n).then(|| n_id),
+        flows_to_load: node_does_load(n).then(|| n_id),
+        flows_to_pos_offset: node_does_pos_offset(n).then(|| n_id),
+        flows_to_neg_offset: node_does_neg_offset(n).then(|| n_id)
+    }
+}
+
+fn create_flow_info (g: &Graph) -> HashMap<NodeId,GraphTraverseInfo> {
+    let mut f = HashMap::from_iter(
+        g.nodes.iter_enumerated()
+        .map(|(idx,node)| (idx,init_traverse_info(idx,node)))
+    );
+    for (n_id,node) in g.nodes.iter_enumerated().rev(){ 
+        let cur : GraphTraverseInfo = *(f.get(&n_id).unwrap());
+        if let Some(p_id) = node.source {
+            let parent = f.get_mut(&p_id).unwrap();
+            parent.last_descendent = cmp::max(cur.last_descendent,parent.last_descendent);
+            parent.flows_to_load = parent.flows_to_load.or(cur.flows_to_load);
+            parent.flows_to_store = parent.flows_to_store.or(cur.flows_to_store);
+            parent.flows_to_pos_offset = parent.flows_to_pos_offset.or(cur.flows_to_pos_offset);
+            parent.flows_to_neg_offset = parent.flows_to_neg_offset.or(cur.flows_to_neg_offset);
+        }
+    }
+    f
+}
+
+fn collect_children (g: &Graph) -> HashMap<NodeId,Vec<NodeId>> {
+    let mut m = HashMap::new();
+    for (par,chi) in g.nodes.iter_enumerated().filter_map(|(idx,node)| Some((node.source?,idx))){
+        m.entry(par).or_insert_with(Vec::new).push(chi)
+    };
+    for (par,chi) in g.nodes.iter_enumerated(){
+        m.try_insert(par,Vec::new());
```

Comment:
i think this loop should happen first, and the second loop should just push to the existing vector